tests/unit/s2n_server_psk_extension_test.c

```diff
@@ -0,0 +1,514 @@
+
+            struct s2n_psk *shared_psk = NULL;
+            EXPECT_OK(setup_order_mismatch_psks(server_conn, client_conn, &shared_psk));
+            EXPECT_NOT_NULL(shared_psk);
+
+            uint16_t shared_psk_wire_index = 0;
+
+            for (size_t i = 0; i < client_conn->psk_params.psk_list.len; i++) {
+                struct s2n_psk *psk = NULL;
+                EXPECT_OK(s2n_array_get(&client_conn->psk_params.psk_list, i, (void**)&psk));
+                if (psk == shared_psk) {
+                    shared_psk_wire_index = i;
+                }
+            }
+
+            /* Valid chosen psk wire index set */
+            server_conn->psk_params.chosen_psk_wire_index = shared_psk_wire_index;
+            EXPECT_SUCCESS(s2n_server_psk_extension.send(server_conn, &out));
+
+            uint16_t chosen_psk_wire_index;
+            EXPECT_SUCCESS(s2n_stuffer_read_uint16(&out, &chosen_psk_wire_index));
+            EXPECT_EQUAL(chosen_psk_wire_index, server_conn->psk_params.chosen_psk_wire_index);
+            EXPECT_TRUE(chosen_psk_wire_index < client_conn->psk_params.psk_list.len);
+
+            EXPECT_SUCCESS(s2n_connection_free(server_conn));
+            EXPECT_SUCCESS(s2n_connection_free(client_conn));
+            EXPECT_SUCCESS(s2n_stuffer_free(&out));
+        }
+    }
+
+    /* Test: s2n_server_psk_recv */
+    {
+        s2n_extension_type_id key_share_ext_id;
+        EXPECT_SUCCESS(s2n_extension_supported_iana_value_to_id(TLS_EXTENSION_KEY_SHARE, &key_share_ext_id));
+        /* Test s2n_server_psk_recv for invalid TLS versions <= TLS1.2 */
+        {
+            struct s2n_stuffer out = { 0 };
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&out, 0));
+
+            struct s2n_connection *conn = NULL;
+            EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+
+            EXPECT_OK(setup_psks(conn));
+
+            /* The keyshare extensions need to be present, as s2n currently only
+             * supports pre-shared keys in (EC)DHE key exchange mode.
+             */
+            conn->psk_params.psk_ke_mode = S2N_PSK_DHE_KE;
+            S2N_CBIT_SET(conn->extension_requests_received, key_share_ext_id);
+
+
+            uint16_t chosen_psk_wire_index = 1;
+            /* Incorrect protocol version */
+            conn->actual_protocol_version = S2N_TLS12;
+            EXPECT_SUCCESS(s2n_stuffer_write_uint16(&out, chosen_psk_wire_index));
+
+            EXPECT_NULL(conn->psk_params.chosen_psk);
+            EXPECT_SUCCESS(s2n_server_psk_extension.recv(conn, &out));
+            EXPECT_NULL(conn->psk_params.chosen_psk);
+
+            /* Validate that the client list of PSKs is not wiped */
+            for (size_t i = 0; i < conn->psk_params.psk_list.len; i++) {
+                struct s2n_psk *psk = NULL;
+                EXPECT_OK(s2n_array_get(&conn->psk_params.psk_list, i, (void**)&psk));
+                EXPECT_OK(validate_psk_is_not_wiped(psk));
+            }
```

Comment:
Understood, the only reason I included this assertion is because our current logic *should*  leave the PSKs not wiped and If the PSKs were wiped there was a problem in the logic flow.